src/ray/object_manager/object_manager.cc

```diff
@@ -400,6 +404,51 @@ void ObjectManager::HandleReceiveFinished(const ObjectID &object_id,
   profile_events_.push_back(profile_event);
 }
 
+void PushManager::StartPush(const UniqueID &push_id, int64_t num_chunks,
+                            std::function<void(int64_t)> send_chunk_fn) {
+  RAY_LOG(DEBUG) << "Start push for " << push_id << ", num chunks " << num_chunks;
+  RAY_CHECK(num_chunks > 0);
+  push_info_[push_id] = std::make_pair(num_chunks, send_chunk_fn);
+  next_chunk_id_[push_id] = 0;
+  chunks_remaining_ += num_chunks;
+  ScheduleRemainingPushes();
+  RAY_CHECK(push_info_.size() == next_chunk_id_.size());
+}
+
+void PushManager::OnChunkComplete() {
+  chunks_in_flight_ -= 1;
+  ScheduleRemainingPushes();
+}
+
+// TODO(ekl) currently chunks are sent in somewhat arbitrary order, we might want to
+// prioritize chunk sends in FIFO or round robin according to push id.
+void PushManager::ScheduleRemainingPushes() {
+  auto it = push_info_.begin();
+  while (it != push_info_.end() && chunks_in_flight_ < max_chunks_in_flight_) {
+    auto push_id = it->first;
+    auto max_chunks = it->second.first;
+    auto send_chunk_fn = it->second.second;
+
+    // Send the next chunk for this push.
+    send_chunk_fn(next_chunk_id_[push_id]);
+    chunks_in_flight_ += 1;
+    chunks_remaining_ -= 1;
+    RAY_LOG(DEBUG) << "Sending chunk " << next_chunk_id_[push_id] << " of " << max_chunks
+                   << " for push " << push_id << ", chunks in flight "
+                   << NumChunksInFlight() << " / " << max_chunks_in_flight_
+                   << " max, remaining chunks: " << NumChunksRemaining();
+
+    // It is the last chunk and we don't need to track it any more.
+    if (++next_chunk_id_[push_id] >= max_chunks) {
+      next_chunk_id_.erase(push_id);
+      push_info_.erase(it++);
+      RAY_LOG(DEBUG) << "Push for " << push_id
+                     << " completed, remaining: " << NumPushesInFlight();
+    }
+    // Note that we don't advance the iterator until done with the push.
```

Comment:
Doesn't this mean there can be starvation? Is `flat_hash_map` keeping the insertion order?

Reply:
Good point. Let me switch this to randomized round robin.